GitUI/Avatars/CustomAvatarProvider.cs

```diff
@@ -0,0 +1,102 @@
+            _subProvider = subProvider ?? throw new ArgumentNullException(nameof(subProvider));
+        }
+
+        /// <inheritdoc/>
+        public async Task<Image> GetAvatarAsync([NotNull] string email, string name, int imageSize)
+        {
+            var templateData = new UriTemplateData(email, name, imageSize);
+
+            foreach (var provider in _subProvider)
+            {
+                var avatar = provider switch
+                {
+                    UriTemplateResolver r => await _downloader.DownloadImageAsync(r.ResolveTemplate(templateData)),
+                    _ => await provider.GetAvatarAsync(email, name, imageSize),
+                };
+
+                if (avatar != null)
+                {
+                    return avatar;
+                }
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Parses a custom avatar template string and creates an <see cref="IAvatarProvider"/> from it.
+        /// </summary>
+        /// <param name="customProviderTemplates">The custom avatar provider template.</param>
+        /// <param name="downloader">The downloader that is used to download avatar images.</param>
+        /// <returns>Returns the <see cref="IAvatarProvider"/> described by the template.</returns>
+        public static IAvatarProvider ParseTemplateString(string customProviderTemplates, [NotNull] IAvatarDownloader downloader)
+        {
+            if (downloader is null)
+            {
+                throw new ArgumentNullException(nameof(downloader));
+            }
+
+            var providerParts = customProviderTemplates
+                .Split(';')
+                .Select(p => p.Trim())
+                .Select(p => FromTemplateSegment(downloader, p))
+                .Where(p => p != null)
+                .ToArray();
+
+            // We can't use the chain provider here, because some returned providers (namely UriTemplateResolvers)
+            // don't actually fulfill the interface contract of IAvatarProvider. UriTemplateResolver is a special
+            // case that exists to prevent variables like hashes from being evaluated/calculated multiple times.
+
+            return new CustomAvatarProvider(downloader, providerParts);
+        }
+
+        /// <summary>
+        /// Parses a single template segment.
+        /// </summary>
+        private static IAvatarProvider FromTemplateSegment(IAvatarDownloader downloader, string providerTemplate)
+        {
+            // if the segment is a tag like "<Demo>", we extract the name
+            // and try to parse it as an AvatarProvider enum.
+            if (providerTemplate.StartsWith("<") && providerTemplate.EndsWith(">"))
+            {
+                var providerName = providerTemplate.Substring(1, providerTemplate.Length - 2);
+
+                if (Enum.TryParse<AvatarProvider>(providerName, true, out var provider)
+                    && provider != AvatarProvider.Custom)
```

Comment:
`<None>;<Default>;***;***`

Will this be the correct pattern?

`<None>` - Do nothing.

Or do you want to use `<None> + fallback + <Default> + fallback + *** + fallback `?

Why not explicitly list how it works now, but it is necessary to use a hidden combination of providers?

AvatarProvider.None - should be really none without fallbacks.
AvatarProvider.Github
AvatarProvider.Gravatar - i guess Gravatar parameters should be not in fallbacks
AvatarProvider.Custom - should be use `<None>, <Github>, <Gravatar|Identicon>, ***`.

Fallbacks must explicitly contain all used providers..

Consider using JSON. It may simplify parsing.